### PR TITLE
SDIT-860 Use native query for offender sentence details

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java [hunk 1]
```diff
@@ -21,17 +20,19 @@
...
-        return findOffendersByNomsId(nomsId, PageRequest.of(0,1)).stream().findFirst();
+    default Optional<Offender> findOffenderByNomsId(String nomsId) {
+        return findOffendersByNomsId(nomsId, PageRequest.of(0, 1)).stream().findFirst();
     }
 
-    @EntityGraph(value = "offender-with-sentences")
-    Optional<Offender> findFirstWithSentencesByNomsId(String nomsId);
+    @Query("select o from Offender o left join fetch o.bookings b left join fetch b.releaseDetail r " +
+        "WHERE o.nomsId = :nomsId order by b.bookingSequence asc limit 1")
+    Optional<Offender> findOffendersWithReleaseDetailByNomsId(@Param("nomsId") String nomsId);
```
Comment (karenmillermoj) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java:35]: Really struggled with this one - the entity graph was returning too much information, so had to revert to just grabbing the details using native sq.  The limit 1 is just to mimic the previous call that was using .findFirst();(line 30).
This however does show _HHH90003004: firstResult/maxResults specified with collection fetch; applying in memory_ (as does the findFirst method)

Comment (andymarke) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java:35]: So essentially you want the latest bookingId? So I suspect you want to to say 
"select o from Offender o left join fetch o.bookings b left join fetch b.releaseDetail r  where o.nomisId = :nomisId and b.bookingSequence = 1"

Would that always return a single row? Maybe try it in SQL first

**Review by andymarke — APPROVED** [review]
LGTM but maybe take a look at comment

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java (hunk 1 above)
Comment (karenmillermoj) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java:35]: ah ok - sounds reasonable - didn't think about that option... will try

**Review by andymarke — APPROVED** [review]
lgtm

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java [hunk 2]
```diff
@@ -21,17 +20,18 @@
...
-    default Optional<Offender> findOffenderByNomsId(String nomsId){
-        return findOffendersByNomsId(nomsId, PageRequest.of(0,1)).stream().findFirst();
+    default Optional<Offender> findOffenderByNomsId(String nomsId) {
+        return findOffendersByNomsId(nomsId, PageRequest.of(0, 1)).stream().findFirst();
     }
 
-    @EntityGraph(value = "offender-with-sentences")
-    Optional<Offender> findFirstWithSentencesByNomsId(String nomsId);
+    @Query("select o from Offender o left join fetch o.bookings b WHERE o.nomsId = :nomsId and b.bookingSequence = 1")
+    Optional<Offender> findOffendersWithLatestBookingByNomsId(@Param("nomsId") String nomsId);
```
[the remark this answers is not in the document]
Comment (karenmillermoj) [reply src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/repository/OffenderRepository.java:34]: ha ha I did that initially and then changed it to match others.  Will update

**Review by andymarke — APPROVED** [review]
lgtm

**Pull request merged**